### PR TITLE
AP_RangeFinder: add serial driver for JRE

### PR DESCRIPTION
**Pull request by jfbblue0922**
JAE is creating a new radio altimeter(JRE).
This is the PR of that serial driver.

Thank you.

### CODE REVIEW

File: libraries/AP_RangeFinder/AP_RangeFinder_JRE_Serial.cpp [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+#define FRAME_HEADER_2   0x41  // 'A'
+
+#define DIST_MAX_CM 50000
+
+#define CCITT_INITPARA   0xffff
+#define CCITT_POLYNOMIAL 0x8408
+#define CCITT_OUTPARA    0xffff
+
+/* CRC16CCITT */
+static uint16_t CalcCRC16CCITT(uint8_t *cbuffer, uint16_t csize)
```
Comment (muramura) [inline libraries/AP_RangeFinder/AP_RangeFinder_JRE_Serial.cpp:32]: There are already two in CHIBIOS.
I think it would be better to add them below and share them with you for future use.

libraries/AP_Math/crc.cpp

File: libraries/AP_RangeFinder/AP_RangeFinder_JRE_Serial.cpp [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+            } else {
+                crc >>= 1;
+            }
+        }
+    }
+    crc = crc ^ CCITT_OUTPARA;
+    return crc;
+}
+
+bool AP_RangeFinder_JRE_Serial::get_reading(float &reading_m)
```
Comment (muramura) [inline libraries/AP_RangeFinder/AP_RangeFinder_JRE_Serial.cpp:50]: Can you describe the frame configuration as there does not seem to be a datasheet published.
BYTE 
0 : FRAME_HEADER_1 'R'
1: FRAME_HEADER_2 'A'
BYTE 0 : FRAME_HEADER_1 'R'
: :

File: libraries/AP_RangeFinder/AP_RangeFinder_JRE_Serial.cpp [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "AP_RangeFinder_JRE_Serial.h"
+
+#if AP_RANGEFINDER_JRE_SERIAL_ENABLED
+
+#include <GCS_MAVLink/GCS.h>
+
+#define FRAME_HEADER_1   0x52  // 'R'
```
Comment (muramura) [inline libraries/AP_RangeFinder/AP_RangeFinder_JRE_Serial.cpp:22]: I think it would be better to define 'R' and 'A' if they are indicated by comments.

File: libraries/AP_RangeFinder/AP_RangeFinder_Params.cpp [hunk 4]
```diff
@@ -15,7 +15,7 @@ const AP_Param::GroupInfo AP_RangeFinder_Params::var_info[] = {
...
-    // @Values: 0:None,1:Analog,2:MaxbotixI2C,3:LidarLite-I2C,5:PWM,6:BBB-PRU,7:LightWareI2C,8:LightWareSerial,9:Bebop,10:MAVLink,11:USD1_Serial,12:LeddarOne,13:MaxbotixSerial,14:TeraRangerI2C,15:LidarLiteV3-I2C,16:VL53L0X or VL53L1X,17:NMEA,18:WASP-LRF,19:BenewakeTF02,20:Benewake-Serial,21:LidarLightV3HP,22:PWM,23:BlueRoboticsPing,24:DroneCAN,25:BenewakeTFminiPlus-I2C,26:LanbaoPSK-CM8JL65-CC5,27:BenewakeTF03,28:VL53L1X-ShortRange,29:LeddarVu8-Serial,30:HC-SR04,31:GYUS42v2,32:MSP,33:USD1_CAN,34:Benewake_CAN,35:TeraRangerSerial,36:Lua_Scripting,37:NoopLoop_TOFSense, 100:SITL
+    // @Values: 0:None,1:Analog,2:MaxbotixI2C,3:LidarLite-I2C,5:PWM,6:BBB-PRU,7:LightWareI2C,8:LightWareSerial,9:Bebop,10:MAVLink,11:USD1_Serial,12:LeddarOne,13:MaxbotixSerial,14:TeraRangerI2C,15:LidarLiteV3-I2C,16:VL53L0X or VL53L1X,17:NMEA,18:WASP-LRF,19:BenewakeTF02,20:Benewake-Serial,21:LidarLightV3HP,22:PWM,23:BlueRoboticsPing,24:DroneCAN,25:BenewakeTFminiPlus-I2C,26:LanbaoPSK-CM8JL65-CC5,27:BenewakeTF03,28:VL53L1X-ShortRange,29:LeddarVu8-Serial,30:HC-SR04,31:GYUS42v2,32:MSP,33:USD1_CAN,34:Benewake_CAN,35:TeraRangerSerial,36:Lua_Scripting,37:NoopLoop_TOFSense,38:JRE_Serial, 100:SITL
```
Comment (muramura) [inline libraries/AP_RangeFinder/AP_RangeFinder_Params.cpp:18]: ", 100:XXXX" with a comma followed by whitesbase.
I think it is unnecessary.

File: libraries/AP_RangeFinder/AP_RangeFinder_JRE_Serial.h [hunk 5]
```diff
@@ -0,0 +1,53 @@
+#pragma once
+
+#include "AP_RangeFinder.h"
+#include "AP_RangeFinder_Backend_Serial.h"
+
+#ifndef AP_RANGEFINDER_JRE_SERIAL_ENABLED
```
Comment (rmackay9) [inline libraries/AP_RangeFinder/AP_RangeFinder_JRE_Serial.h:6]: It's great to see this driver!

I think we should move this _ENABLED definition to AP_RangeFinder_config.h.  The pattern has changed just in the last two weeks but still.. it would be good to stick with the new pattern.

File: libraries/AP_RangeFinder/AP_RangeFinder_JRE_Serial.cpp (hunk 1 above)
Comment (rmackay9) [reply libraries/AP_RangeFinder/AP_RangeFinder_JRE_Serial.cpp:32]: I agree withe @muramura that **if possible** we should re-use one of the existing crc methods ([see here](https://github.com/ArduPilot/ardupilot/blob/master/libraries/AP_Math/crc.h#L22))

File: libraries/AP_RangeFinder/AP_RangeFinder_JRE_Serial.cpp (hunk 3 above)
Comment (jfbblue0922) [reply libraries/AP_RangeFinder/AP_RangeFinder_JRE_Serial.cpp:22]: I corrected definition of "R" and "A".

File: libraries/AP_RangeFinder/AP_RangeFinder_Params.cpp (hunk 4 above)
Comment (jfbblue0922) [reply libraries/AP_RangeFinder/AP_RangeFinder_Params.cpp:18]: I corrected it.

File: libraries/AP_RangeFinder/AP_RangeFinder_JRE_Serial.h (hunk 5 above)
Comment (jfbblue0922) [reply libraries/AP_RangeFinder/AP_RangeFinder_JRE_Serial.h:6]: I moved this _ENABLED definition to AP_RangeFinder_config.h.

File: libraries/AP_RangeFinder/AP_RangeFinder_JRE_Serial.cpp (hunk 2 above)
Comment (jfbblue0922) [reply libraries/AP_RangeFinder/AP_RangeFinder_JRE_Serial.cpp:50]: The data format is shown below.
Fig1

File: libraries/AP_RangeFinder/AP_RangeFinder_JRE_Serial.cpp (hunk 1 above)
Comment (jfbblue0922) [reply libraries/AP_RangeFinder/AP_RangeFinder_JRE_Serial.cpp]: I moved crc calculation to libraries/AP_Math/crc.cpp.

File: libraries/AP_Math/crc.cpp [hunk 6]
```diff
@@ -374,6 +374,53 @@ uint16_t crc16_ccitt(const uint8_t *buf, uint32_t len, uint16_t crc)
     return crc;
 }
 
+/* right shift table */
```
Comment (rmackay9) [inline libraries/AP_Math/crc.cpp:377]: @jfbblue0922,

Great, thanks for this.  We should actually move this to a separate commit prefixed with "AP_Math:".

Comment (jfbblue0922) [reply libraries/AP_Math/crc.cpp:377]: @rmackay9 

thank you.
I have never split commits.
Would it be possible to revert the fix with the 'git revert' command and then make another commit prefixed with AP_Math and AP_RangeFinder?

Comment (rmackay9) [reply libraries/AP_Math/crc.cpp:377]: @jfbblue0922,

If you're using Linux/Ubuntu then type "git gui" into the terminal and then a screen like below should pop-up.  "Amend Last Commit", then uncheck the two AP_Math changes, then press Commit again.  Then create a new commit with just the AP_Math changes by staging and then committing them.  Then do a forced push ("git push -f").
image

If you need my help after this though I can do it for you.

File: libraries/AP_RangeFinder/AP_RangeFinder_JRE_Serial.cpp [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+                if ((((crc>>8) & 0xff) == data_buff[15]) && ((crc & 0xff) == data_buff[14])) {
+                    // status check
+                    if (data_buff[13] & 0x02) { // NTRK
+                        no_signal = true;
+                        reading_m = MIN(MAX(DIST_MAX_CM, distance_cm_max), UINT16_MAX) * 0.01f;
+                    } else { // UPDATE DATA
+                        no_signal = false;
+                        reading_cm = data_buff[4] * 256 + data_buff[5];
+                        if (reading_cm < distance_cm_max) {
+                            reading_m = reading_cm * 0.01f;
```
Comment (rmackay9) [inline libraries/AP_RangeFinder/AP_RangeFinder_JRE_Serial.cpp:93]: It looks like this function will only ever return the last distance that it is in the buffer at the time this get_reading() is run.  Normally our rangefinder drivers return the average of all readings.
I think this also means that if it getse two readings and the last one is "NTRK" it will return no_signal instead of the earlier good reading.

File: libraries/AP_RangeFinder/AP_RangeFinder_JRE_Serial.cpp [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+                        } else {
+                            reading_m = distance_cm_max * 0.01f;
```
Comment (rmackay9) [inline libraries/AP_RangeFinder/AP_RangeFinder_JRE_Serial.cpp:95]: I think that most other drivers will return a distance further than the max but then the range finders status will be changed to out-of-range-high.  It would be good to check the states change correctly.

File: libraries/AP_RangeFinder/AP_RangeFinder_JRE_Serial.cpp [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+            // header data search
+            for (; read_buff_idx<num_read; read_buff_idx++) {
+                if (read_buff[read_buff_idx] == FRAME_HEADER_1) {
+                    data_buff[0] = FRAME_HEADER_1;
+                    data_buff_idx = 1; // next data_buff
+                    if (read_buff_idx >= num_read - 1) { // read end byte
+                        if (count > 0) {
+                            return true;
+                        }
+                        return false;      // next packet to header second byte
```
Comment (rmackay9) [inline libraries/AP_RangeFinder/AP_RangeFinder_JRE_Serial.cpp:57]: I think it might be good to restructure this state machine so that we can avoid the duplicate, "(read_buff_idx >= num_read - 1)" check.

**Comment by rmackay9** [comment]
@jfbblue0922,

If possible it would be good to squash the first 3 commits together.  I do this from the terminal by typing, "git rebase -i "HEAD~10".  This opens up an editor (vim for me) and you can then change "push" to "s" and then save and exit the editor (":ws")
image

Then the editor will show the new single commit's comments like below.  I delete some of the lower lines (press "d" a few times) and save again.
image

File: libraries/AP_RangeFinder/AP_RangeFinder_JRE_Serial.cpp (hunk 8 above)
Comment (jfbblue0922) [reply libraries/AP_RangeFinder/AP_RangeFinder_JRE_Serial.cpp:83]: I corrected it.

File: libraries/AP_RangeFinder/AP_RangeFinder_JRE_Serial.cpp (hunk 9 above)
Comment (jfbblue0922) [reply libraries/AP_RangeFinder/AP_RangeFinder_JRE_Serial.cpp]: I corrected it.

**Comment by jfbblue0922** [comment]
@rmackay9 

I think I was able to squash.
thank you.

File: libraries/AP_RangeFinder/AP_RangeFinder_JRE_Serial.cpp [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+            } else {
+                data_buff_idx = 0;    // data index clear
+            }
+            read_buff_idx++;          // next read_buff
+
+        } else { // data set
+            if (data_buff_idx >= ARRAY_SIZE(data_buff)) {  // 1 data set complete
+                // crc check
+                uint16_t crc = crc16_ccitt_r(data_buff, ARRAY_SIZE(data_buff) - 2, 0xffff, 0xffff);
+                if ((((crc>>8) & 0xff) == data_buff[15]) && ((crc & 0xff) == data_buff[14])) {
```
Comment (rmackay9) [inline libraries/AP_RangeFinder/AP_RangeFinder_JRE_Serial.cpp:74]: Sorry for the drip feed of change request.. but you could use the UINT16_VALUE macro both here and a few lines below to more easily combine the two types into a uint16_t.

File: libraries/AP_RangeFinder/AP_RangeFinder_JRE_Serial.cpp [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+                    // status check
+                    if (data_buff[13] & 0x02) { // NTRK
+                        invalid_count++;
+                    } else { // TRK
+                        reading_cm = data_buff[4] * 256 + data_buff[5];
+                        if (reading_cm < distance_cm_max) {
```
Comment (rmackay9) [inline libraries/AP_RangeFinder/AP_RangeFinder_JRE_Serial.cpp:80]: I think we should remove this "reading_cm < distance_cm_max" check.  Imagine this scenarios:

- max distance is set to 50m
- two distances arrive, 75m and 45m
- 75m will be shortened to 50m, the average will become 47.5m which will be considered valid and will be used by the autopilot (for surface tracking for example)

Instead what we should do is average 75 and 45 to arrive at 60m which will then be determined (by higher level code) to be out-of-range-high and the value will not be used.

File: libraries/AP_RangeFinder/AP_RangeFinder_JRE_Serial.h [hunk 12]
```diff
@@ -0,0 +1,49 @@
+#pragma once
+
+#include "AP_RangeFinder.h"
+#include "AP_RangeFinder_Backend_Serial.h"
+
+#if AP_RANGEFINDER_JRE_SERIAL_ENABLED
+
+#define DATA_LENGTH 16
```
Comment (rmackay9) [inline libraries/AP_RangeFinder/AP_RangeFinder_JRE_Serial.h:8]: I think we should prefix this definition with something like AP_RANGEFINDER_JRE_ (or similar) because this definition is in the global scope and because it is quite generic it could be re-used by another driver.  There are no conflicts at the moment but normally we add a prefix I think.

File: libraries/AP_RangeFinder/AP_RangeFinder_JRE_Serial.cpp (hunk 2 above)
Comment (rmackay9) [reply libraries/AP_RangeFinder/AP_RangeFinder_JRE_Serial.cpp:27]: If possible I think we should add this packet structure into the comments at the top of the .cpp or .h file.  It's important for future developers to know the packet format in case they need to modify the code a bit.

File: libraries/AP_RangeFinder/AP_RangeFinder_JRE_Serial.cpp (hunk 10 above)
Comment (jfbblue0922) [reply libraries/AP_RangeFinder/AP_RangeFinder_JRE_Serial.cpp]: I corrected it.

File: libraries/AP_RangeFinder/AP_RangeFinder_JRE_Serial.cpp (hunk 11 above)
Comment (jfbblue0922) [reply libraries/AP_RangeFinder/AP_RangeFinder_JRE_Serial.cpp]: I corrected it.

File: libraries/AP_RangeFinder/AP_RangeFinder_JRE_Serial.h (hunk 12 above)
Comment (jfbblue0922) [reply libraries/AP_RangeFinder/AP_RangeFinder_JRE_Serial.h:10]: I corrected it.
thank you.

**Comment by rmackay9** [comment]
Thanks for the updates.  I've added the DevCallTopic label so this will be on the agenda for review at next week's Dev call (Tue 8am Japan time).

**Review by peterbarker — CHANGES_REQUESTED** [review]
Very partial review

**Comment by rmackay9** [comment]
We discussed on the dev call and decided that @peterbarker will write a simulator and I will rework the state machine to remove the while and make it more like our other rangefinders.  We suspect that the current code is safe but it is a bit "fragile" (as we say) meaning that the while loop means that future changes could cause an infinite loop.. it's just not a pattern we want in sensor drivers like this.

File: libraries/AP_Math/crc.cpp [hunk 13]
```diff
@@ -374,6 +374,53 @@ uint16_t crc16_ccitt(const uint8_t *buf, uint32_t len, uint16_t crc)
...
+static const uint16_t crc16tab_r[256] = {
```
Comment (magicrub) [inline libraries/AP_Math/crc.cpp:378]: Which CCITT is this? https://crccalc.com/ There's a 99.9% chance it's already implemented here

File: libraries/AP_RangeFinder/AP_RangeFinder.cpp [hunk 14]
```diff
@@ -539,6 +540,11 @@ void RangeFinder::detect_instance(uint8_t instance, uint8_t& serial_instance)
     case Type::NoopLoop_P:
 #if AP_RANGEFINDER_NOOPLOOP_ENABLED
         serial_create_fn = AP_RangeFinder_NoopLoop::create;
+#endif
+        break;
+    case Type::JRE_Serial:
+#if AP_RANGEFINDER_JRE_SERIAL_ENABLED
```
[the remark this answers is not in the document]
Comment (jfbblue0922) [reply libraries/AP_RangeFinder/AP_RangeFinder.cpp:546]: I corrected it.
but other device's code is inside of the case declaration.

File: libraries/AP_RangeFinder/AP_RangeFinder_JRE_Serial.cpp [hunk 15]
```diff
@@ -0,0 +1,105 @@
...
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "AP_RangeFinder_JRE_Serial.h"
+
+#if AP_RANGEFINDER_JRE_SERIAL_ENABLED
```
[the remark this answers is not in the document]
Comment (jfbblue0922) [reply libraries/AP_RangeFinder/AP_RangeFinder_JRE_Serial.cpp:18]: I corrected it.

File: libraries/AP_RangeFinder/AP_RangeFinder_JRE_Serial.cpp [hunk 16]
```diff
@@ -0,0 +1,105 @@
...
+{
+    // uart instance check
+    if (uart == nullptr) {
+        return false;  // not update
+    }
+    uint32_t n = uart->available();
+    if (n == 0) {
+        return false;
+    }
+
```
[the remark this answers is not in the document]
Comment (jfbblue0922) [reply libraries/AP_RangeFinder/AP_RangeFinder_JRE_Serial.cpp:38]: I corrected it.

File: libraries/AP_RangeFinder/AP_RangeFinder_JRE_Serial.cpp (hunk 2 above)
Comment (jfbblue0922) [reply libraries/AP_RangeFinder/AP_RangeFinder_JRE_Serial.cpp:29]: Since no datasheet has been published for this device, I will add more details to the header file.

**Comment by jfbblue0922** [comment]
@rmackay9 

Should we fix the while loop?

File: libraries/AP_RangeFinder/AP_RangeFinder_JRE_Serial.cpp [hunk 17]
```diff
@@ -0,0 +1,102 @@
...
+    uint16_t valid_count = 0;   // number of valid readings
+    uint16_t invalid_count = 0; // number of invalid readings
+    float sum = 0;
+    // max distance the sensor can reliably measure - read from parameters
+    const int16_t distance_cm_max = max_distance_cm();
+
+    // buffer read
+    const ssize_t num_read = uart->read(read_buff, ARRAY_SIZE(read_buff));
+    read_buff_idx = 0; // read_buff start data index
+    while (read_buff_idx < num_read) {
```
Comment (muramura) [inline libraries/AP_RangeFinder/AP_RangeFinder_JRE_Serial.cpp:45]: I agree with Randy's comments.

If READ_BUFF_IDX is incremented in the condition of the WHILE statement, the loop will not be infinite.
If there is an increment leak during processing within the loop, the loop will go around indefinitely.

In other range finders, the index is updated under the condition of the WHILE statement.

**Comment by rmackay9** [comment]
@jfbblue0922,

Would it be possible for me to get one of these sensors?  PeterB has written a simulator for this device but having the real device would allow me to test it as part of the restructure.  In answer to your question, I think we should restructure the while loop.  I'm happy to do that.

File: libraries/AP_RangeFinder/AP_RangeFinder_JRE_Serial.cpp (hunk 17 above)
Comment (jfbblue0922) [reply libraries/AP_RangeFinder/AP_RangeFinder_JRE_Serial.cpp]: @muramura 

I changed the 'while loop' to a 'for loop'.

and
I added the codes to adapt additional protocols of this sensor.
See header file (AP_RangeFinder_JRE_Serial.h).

**Comment by jfbblue0922** [comment]
@rmackay9 

Regarding the while loop, We have already pushed the code using a for loop.

We can lend a sensor of prototyping one.
It's communication protocol is same.

Also, could you please write a loan agreement like last time?

thank you.

**Review by Hwurzburg — DISMISSED** [review]
this also needs the ifdefs around the backend, inclusion in the config file, addition to the build_options and , possibly, the extract features (unless its regex covers it) scripts

look at hoe the other backends are done for these

**Comment by jfbblue0922** [comment]
@Hwurzburg 

Thank you for your point out.
I modified the script and committed it.
Is it correct?

**Comment by rmackay9** [comment]
@jfbblue0922, sure, i can do the loan agreement.  I guess you'll email it along or do I just use the previous template?

**Comment by jfbblue0922** [comment]
@Hwurzburg 

I corrected it.
Thank you.

**Comment by jfbblue0922** [comment]
@rmackay9 

We will send it to you soon.
Thank you.

**Comment by jfbblue0922** [comment]
@rmackay9 

We just sent email with the document to you.
Thank you.

**Comment by jfbblue0922** [comment]
@rmackay9 

We have resent the email.
If the transmission fails, we will try with another method.

Thank you.
